crates/transaction-emitter-lib/src/args.rs

```diff
@@ -69,13 +69,13 @@ impl Default for TransactionType {
 
 #[derive(Clone, Debug, Default, Deserialize, Parser, Serialize)]
 pub struct EmitArgs {
-    #[clap(long, default_value = "15")]
-    pub accounts_per_client: usize,
+    #[clap(long, default_value = "5000")]
+    /// Number of transactions outstanding in mempool - this is needed to ensure that the emitter
+    /// is producing enough load to get the highest TPS in the system. Typically this should be
+    /// configured to be ~4x of the max achievable TPS.
+    pub mempool_backlog: u64,
 
-    #[clap(long)]
-    pub workers_per_ac: Option<usize>,
-
-    #[clap(long, default_value = "0")]
+    #[clap(long, default_value = "0", requires = "burst")]
```

Comment:
We still have burst mode support (although I am tempted to remove it at some point of time) - the `wait_millis` is only applicable in case of `burst` mode now.

Reply:
we have a few discussions on fix TPS (burst mode) in the transaction emitter, I still see open loop is a better way to load test the system than close loop. we're interested in learning how the system behaves with controlled incoming TPS to the system. also cc @ikabiljo @bchocho for discussion

Reply:
I feel like the sequence numbers issue is affecting our decisions here.

If we were to ignore sequence numbers (i.e. we had a system where each transaction is independent, and we can send it at any point), how would we want load test to look like?

I think we would want to have a way to achieve a given fixed TPS (that is purely time based, not based on response from the clients), and have a way to have load test modify that TPS dynamically. 
We define some "health" metric of the system, and increase TPS until that health metric is too bad, and decrease it until it is good enough.

What can we make the health metric be? 
I think latency of the transaction being committed, and drop rate of the transaction is probably a good metric. So we can sample a set of transactions, and after a fixed interval (30s?) check how many of them succeeded. And we shoot for 0.1% or something. (i.e. increase TPS if below 0.01% and decrease if above 0.1% for example.

How do we deal with sequence numbers implementation detail? we just rotate accounts often enough. If cost of account creation is 3x the regular transaction, rotating account after every ~20-30 transactions on that account is probably good enough, and small enough that with targeted failure rate - it will not affect the consecutive transactions much.

This will reduce amount of (configuration) complexity txn emitter needs to deal with? 
What do you think?

Reply:
> I still see open loop is a better way to load test the system than close loop. we're interested in learning how the system behaves with controlled incoming TPS to the system.

I am not sure what is the problem we are trying to solve by "open loop" system. After the current changes, the transaction emitter's job is pretty simple and intuitive - all it does is aim to have X number of pending transactions in the mempool - this is very similar to measuring latency as a health check metrics - If the system is fast and commits transactions quickly - then the emitter is going to be more aggressive in submitting the transactions and if the system is slow, the transaction emitter is going to slow down as well - however, the guarantees it provides is that we always have constant load in the backlog to be processed at any point of time. This I am hoping will reduce some of the variables, we used to observe in different experiment runs.

Reply:
the problem open loop solves is to avoid feedback loop caused by the system and have more control on the experiment. I just want to decouple the input to the system vs the performance of the system itself which is much simpler to reason about and experiment with. close loop is unavoidably affected by latencies of the system and number of accounts we use. by purely increasing latency by 1s even the throughput is higher (e.g. quorum store) will result in a lower TPS.

It'd be good to see a plot of latency as Y axis, incoming TPS as X axis to analyze the performance which is also not doable with closed loop.

Reply:
open loop system is really the only way to have a constant TPS. 

Current solution tries to keep the amount of transactions in the system constant, and with it - constantly adjusts the TPS. with that it is unable to keep the fixed TPS for extended period of time and show how system behaves.
If a node falls off, and state sync kicks in, we are going to reduce TPS until it catches up and can handle it again. During epoch change, if things become slower, we are going to temporarily reduce TPS for it to handle it.

We might want to change TPS to find a "max load", but then probably want to fix that load for extended periods of time and see how system performs.

current approach might be useful too, but fixed TPS is something we probably want to have.

Reply:
>  I just want to decouple the input to the system vs the performance of the system itself which is much simpler to reason about and experiment with. 

@zekun000 - This is what the current PR is doing. I think we need to align on what we consider input to the system. In my view the input is pending transaction - we always ensure that the pending transactions are fixed, so that the experiment is more predictable. 


> close loop is unavoidably affected by latencies of the system and number of accounts we use. by purely increasing latency by 1s even the throughput is higher (e.g. quorum store) will result in a lower TPS.

This should not be the case, current configuration guarantees that we always have enough backlog in the system - so even if the latency is increased by changes like Quorum store, we should see higher TPS.  If that's not the case, something wrong with either the emitter or the changes are not really increasing the TPS :). @bchocho - Let's sync up on this if we have some confusion on behavior of forge test for quorum store testing.


> It'd be good to see a plot of latency as Y axis, incoming TPS as X axis to analyze the performance which is also not doable with closed loop

The current system can actually do that with some minor tweaks - basically, we can run multiple experiments by gradually increasing the mempool backlog and you can plot a latency vs TPS graph. The default configuration I have is finding the sweet spot - which aims to achieve the highest throughput possible today with some bounded latency. 

> Current solution tries to keep the amount of transactions in the system constant, and with it - constantly adjusts the TPS. with that it is unable to keep the fixed TPS for extended period of time and show how system behaves.

I think there is some confusion around what we are referring to as TPS.  In my mind, TPS is not transactions submitted per second but the transaction committed per second and the current system should be able to sustain the max possible fixed TPS for an extended period of time. 

Of course, there are other and more fancier ways to do load testing but IMO, what we have is good enough for our requirement ATM 😀.

Reply:
pending transactions is not input to the system, it's metrics from the system that's good for feedback loop, but we want to have the control over pure input too.

@ikabiljo also pointed out we want to see the system behavior with constant load, it may slow down because of epoch change or state sync or anything, but closed loop would hide all those by reducing the load automatically.

we should start a meeting for more effective discussion.

Reply:
one example I just hit is to understand the bls signatures PR, why some runs report 4k TPS and another report 7k. I'd like to see a controlled experiment where both run with fixed incoming transactions and how they behave.

Reply:
Let's take it offline - will setup a quick sync on this.